### PR TITLE
[Product Sharing] Fix Share sheet popover in iPad

### PR DESCRIPTION
**Pull request by selanthiraiyan**
Part of: #9882 

## Description

The share sheet popover is not presented properly in iPad. This PR attempts to fix it by using the [ShareSheet](https://github.com/woocommerce/woocommerce-ios/blob/trunk/WooCommerce/Classes/ViewRelated/ReusableViews/SwiftUI%20Components/ShareSheet.swift) helper.

- For First product view, the popover will be presented only when on iPad and only when AI generation is not enabled. If otherwise, the control will be handed over to `ShareProductCoordinator` which will take over the navigation. 
- For `ProductSharingMessageGenerationView` the share sheet or popover will be presented from SwiftUI without using `SharingHelper`

### Changes 

**Related to First product view**
- Refactor `FirstProductCreatedView` to work with a view model
- Add `ShareProductAIEligibilityChecker` protocol for unit testing. 
- Use [ShareSheet](https://github.com/woocommerce/woocommerce-ios/blob/trunk/WooCommerce/Classes/ViewRelated/ReusableViews/SwiftUI%20Components/ShareSheet.swift) to show popover in iPad when applicable
- Introduce `FirstProductCreatedViewModel` for handling tracking and popover presentation
- Add unit tests to ensure that popover is presented only when on iPad and AI generation is not eligible.

**Related to ProductSharingMessageGenerationView**
- Use [ShareSheet](https://github.com/woocommerce/woocommerce-ios/blob/trunk/WooCommerce/Classes/ViewRelated/ReusableViews/SwiftUI%20Components/ShareSheet.swift) to show share popover/sheet.
- Stop passing message to  `ShareProductCoordinator` and present the share sheet/popover directly from the `ProductSharingMessageGenerationView` SwiftUI view.

**Note**
- [x] To avoid bloating this PR I will add unit tests for `ProductSharingMessageGenerationViewModel` in a future PR - https://github.com/woocommerce/woocommerce-ios/pull/9910

## Testing instructions

**WPCOM store with AI**
1. Create a Free trial store. 
1. Ensure that it is launched and publicly available. (Use store credits to purchase plan)
1. Please test this on iPad (device/simulator is fine)
1. Open the site from the app
1. Create a first product by tapping on "Add your first product" onboarding task
1. You will be presented with a confetti screen after you finish creating and publishing a product
1. Tap on "Share product" and validate that you can generate message using AI
1. Tap "Share" and validate that a share popover is presented
1. Repeat these steps and test on iPhone

**JN site with No AI support**
1. Create a JN site
1. Please test this on iPad (device/simulator is fine)
1. Open the site from the app
1. Create a first product by tapping on "Add your first product" onboarding task
1. You will be presented with a confetti screen after you finish creating and publishing a product
1. Tap on "Share product" and validate that a share popover is presented on iPad.
1. Repeat these steps and test on iPhone

- [x] @selanthiraiyan To test the other share entry points work on iPhone and iPad

## Screenshots
| Store type | iPad | iPhone |
|--------|--------|--------|
| WPCOM | WPCOMiPad |  WPCOMiPhone |
| Self-hosted | SelfHostediPad | SelfHostediPhone |

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/Classes/ViewRelated/Products/Add Product/FirstProductCreated/FirstProductCreatedView.swift [hunk 1]
```diff
@@ -7,23 +7,25 @@ final class FirstProductCreatedHostingController: UIHostingController<FirstProdu
...
-                  let navigationController = self.navigationController,
-                  let productURL = URL(string: product.permalink) else {
+                  let navigationController = self.navigationController else {
                 return
             }
 
-            let shareProductCoordinator = ShareProductCoordinator(productURL: productURL,
-                                                                  productName: product.name,
+            let shareProductCoordinator = ShareProductCoordinator(productURL: viewModel.productURL,
+                                                                  productName: viewModel.productName,
```
Comment (itsmeichigo) [inline WooCommerce/Classes/ViewRelated/Products/Add Product/FirstProductCreated/FirstProductCreatedView.swift]: Nit: we can use the injected `productURL` and `productName` here instead of the `viewModel`'s. That way we can make `launchAISharingFlow` a parameter in the view model's initializer.

File: WooCommerce/Classes/ViewRelated/Products/Add Product/FirstProductCreated/FirstProductCreatedViewModel.swift [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+final class FirstProductCreatedViewModel: ObservableObject {
+    @Published var isSharePopoverPresented = false
+
+    let productURL: URL
+    let productName: String
+    let showShareProductButton: Bool
+    let shareSheet: ShareSheet
+
+    var launchAISharingFlow: (() -> Void)?
```
Comment (itsmeichigo) [inline WooCommerce/Classes/ViewRelated/Products/Add Product/FirstProductCreated/FirstProductCreatedViewModel.swift:12]: Nit: As commented above, we can make this closure private if we inject it in the view model's initializer.

Also - the many different code paths make it hard to follow the logic. Please consider one of these options:
- Rename the closure to something more generic like `startShareCoordinator` and add a comment to explain why it's needed. The reason that we are not using it just for the AI flow but also for the native share sheet on iPhones.
- Keep this closure for just the AI flow, and add an extra property to show the share sheet for iPhones directly from the `FirstProductCreatedView`. This would ensure readability for the logic of the share flows.

File: WooCommerce/Classes/ViewRelated/Products/ShareProduct/ShareProductCoordinator.swift [hunk 3]
```diff
@@ -91,13 +91,13 @@ private extension ShareProductCoordinator {
...
         }
-        let viewModel = ProductSharingMessageGenerationViewModel(siteID: siteID, productName: productName, url: productURL.absoluteString)
-        let controller = ProductSharingMessageGenerationHostingController(viewModel: viewModel) { [weak self] message in
-            self?.navigationController.topmostPresentedViewController.dismiss(animated: true) {
-                self?.presentShareSheet(with: message)
-            }
+        let viewModel = ProductSharingMessageGenerationViewModel(siteID: siteID,
+                                                                 productName: productName,
+                                                                 url: productURL.absoluteString,
+                                                                 onShare: {
```
Comment (itsmeichigo) [inline WooCommerce/Classes/ViewRelated/Products/ShareProduct/ShareProductCoordinator.swift]: Nit: if we are keeping this closure just for analytics, please consider removing it for simplicity because we'll need to add analytics in the view model to track the generation completion/failure anyway.

File: WooCommerce/Classes/ViewRelated/ReusableViews/SwiftUI Components/ShareSheet.swift [hunk 4]
```diff
@@ -36,7 +36,12 @@ private struct ShareSheetView: UIViewControllerRepresentable {
 extension View {
     func shareSheet(isPresented: Binding<Bool>, content: @escaping () -> ShareSheet) -> some View {
         sheet(isPresented: isPresented) {
-            ShareSheetView(shareSheet: content())
+            if #available(iOS 16.0, *) {
+                ShareSheetView(shareSheet: content())
+                    .presentationDetents([.medium, .large])
```
Comment (itsmeichigo) [inline WooCommerce/Classes/ViewRelated/ReusableViews/SwiftUI Components/ShareSheet.swift:41]: 👏

**Review by itsmeichigo — APPROVED** [review]
Great fix!

File: WooCommerce/Classes/ViewRelated/Products/ShareProduct/ShareProductCoordinator.swift (hunk 3 above)
Comment (selanthiraiyan) [reply WooCommerce/Classes/ViewRelated/Products/ShareProduct/ShareProductCoordinator.swift]: I agree. Removed the closure in https://github.com/woocommerce/woocommerce-ios/pull/9901/commits/757c69b64f6897d05bda946fa5e5534d32c985fb

File: WooCommerce/Classes/ViewRelated/Products/Add Product/FirstProductCreated/FirstProductCreatedViewModel.swift (hunk 2 above)
Comment (selanthiraiyan) [reply WooCommerce/Classes/ViewRelated/Products/Add Product/FirstProductCreated/FirstProductCreatedViewModel.swift:12]: Good idea. 💡 

I followed the second option in https://github.com/woocommerce/woocommerce-ios/pull/9901/commits/db01361c4f06265f3fdb530a4c9a7a6c45e4e548 and updated tests in https://github.com/woocommerce/woocommerce-ios/pull/9901/commits/816a6ae06ffd519d254b7cd60b762039781cd1f2

File: WooCommerce/Classes/ViewRelated/Products/Add Product/FirstProductCreated/FirstProductCreatedView.swift (hunk 1 above)
Comment (selanthiraiyan) [reply WooCommerce/Classes/ViewRelated/Products/Add Product/FirstProductCreated/FirstProductCreatedView.swift]: Done in https://github.com/woocommerce/woocommerce-ios/pull/9901/commits/8b3adc6a30b0a7348e4ffc66fa7765484268b713

But I still cannot set `launchAISharingFlow` before initializing the `FirstProductCreatedHostingController`, as we have to read `self` from the closure [here](https://github.com/woocommerce/woocommerce-ios/pull/9901/files#diff-097b3327adf6c0c182f7543892e08b3334297bbe7e9ae55a184fcd6878794de4R18-R19). Kindly let me know if there is another way to achieve this.

Comment (itsmeichigo) [reply WooCommerce/Classes/ViewRelated/Products/Add Product/FirstProductCreated/FirstProductCreatedView.swift]: You are right, since we need reference to `self`, it doesn't seem to be possible.
